### PR TITLE
ARROW-10192: [Python] Always decode inner dictionaries when converting array to Pandas

### PR DESCRIPTION
**Pull request by pitrou**
Fix a crash on conversion of e.g. struct of dictionaries.

### CODE REVIEW

File: cpp/src/arrow/python/arrow_to_pandas.cc [hunk 1]
```diff
@@ -641,29 +659,18 @@ inline Status ConvertStruct(const PandasOptions& options, const ChunkedArray& da
...
-  // array in this function. PyArray_GETITEM returns datetime.datetime for
-  // units second through microsecond but PyLong for nanosecond (because
-  // datetime.datetime does not support nanoseconds).
-  // We force the object conversion to preserve the value of the timezone.
-  // Nanoseconds are returned integers inside of structs.
-  PandasOptions modified_options = options;
-  modified_options.coerce_temporal_nanoseconds = false;
+  options = MakeInnerOptions(std::move(options));
+  // See notes in MakeInnerOptions about timestamp conversion.
+  // Don't blindly convert because
```
Comment (kszucs) [inline cpp/src/arrow/python/arrow_to_pandas.cc:664]: Seems like the second comment line is missing: ` // timestamps in lists are handled differently.`

Comment (pitrou) [reply cpp/src/arrow/python/arrow_to_pandas.cc:664]: Ah, right. Do you want to make the change and push?

Comment (kszucs) [reply cpp/src/arrow/python/arrow_to_pandas.cc]: Yes, updated.

File: python/pyarrow/tests/test_pandas.py [hunk 2]
```diff
@@ -2297,6 +2310,24 @@ def test_from_tuples(self):
...
 
+    def test_struct_of_dictionary(self):
+        names = ['ints', 'strs']
+        children = [pa.array([456, 789, 456]).dictionary_encode(),
+                    pa.array(["foo", "foo", None]).dictionary_encode()]
+        arr = pa.StructArray.from_arrays(children, names=names)
+
+        # Expected a Series of {field name: field value} dicts
+        rows_as_tuples = zip(*(child.to_pylist() for child in children))
+        rows_as_dicts = [dict(zip(names, row)) for row in rows_as_tuples]
```
Comment (kszucs) [inline python/pyarrow/tests/test_pandas.py:2321]: Expanded the nested comprehension to make it a bot more readable.

**Review by kszucs — APPROVED** [review]
Thanks Antoine! Merging on green.

**Pull request closed without merging**